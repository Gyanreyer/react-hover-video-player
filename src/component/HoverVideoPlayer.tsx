import React, { useRef, useImperativeHandle, useEffect, useState } from 'react';

import useSetAdditionalAttributesOnVideo from '../hooks/useSetAdditionalAttributesOnVideo';
import useFormatVideoSrc from '../hooks/useFormatVideoSrc';
import useFormatVideoCaptions from '../hooks/useFormatVideoCaptions';
import useHoverTargetElement from '../hooks/useHoverTargetElement';
import useManageHoverEvents from '../hooks/useManageHoverEvents';

import {
  isVideoElementPaused,
  isVideoElementPlaying,
} from '../utils/videoElementPlaybackStates';

import {
  expandToFillContainerStyle,
  containerSizingStyles,
  pausedOverlayWrapperSizingStyles,
  videoSizingStyles,
} from './HoverVideoPlayer.styles';
import { HoverVideoPlayerProps, VideoSource } from '../HoverVideoPlayer.types';

/**
 * @component HoverVideoPlayer
 * @license MIT
 *
 * @param {HoverVideoPlayerProps} props
 */
export default function HoverVideoPlayer({
  videoSrc,
  videoCaptions = null,
  focused = false,
  disableDefaultEventHandling = false,
  hoverTarget = null,
  onHoverStart = null,
  onHoverEnd = null,
  hoverOverlay = null,
  pausedOverlay = null,
  loadingOverlay = null,
  loadingStateTimeout = 200,
  overlayTransitionDuration = 400,
  playbackStartDelay = 0,
  restartOnPaused = false,
  unloadVideoOnPaused = false,
  playbackRangeStart = null,
  playbackRangeEnd = null,
  muted = true,
  volume = 1,
  loop = true,
  preload = null,
  crossOrigin = null,
  controls = false,
  controlsList = null,
  disableRemotePlayback = true,
  disablePictureInPicture = true,
  style = null,
  hoverOverlayWrapperClassName = null,
  hoverOverlayWrapperStyle = null,
  pausedOverlayWrapperClassName = null,
  pausedOverlayWrapperStyle = null,
  loadingOverlayWrapperClassName = null,
  loadingOverlayWrapperStyle = null,
  videoId = null,
  videoClassName = null,
  videoRef: forwardedVideoRef = null,
  videoStyle = null,
  sizingMode = 'video',
  shouldSuppressPlaybackInterruptedErrors = true,
  ...spreadableProps
}: HoverVideoPlayerProps): JSX.Element {
  // Element refs
  const containerRef = useRef(null);
  const videoRef = useRef<HTMLVideoElement>(null);
  // Forward out local videoRef along to the videoRef prop
  useImperativeHandle(forwardedVideoRef, () => videoRef.current);

  // Effect sets attributes on the video which can't be done via props
  useSetAdditionalAttributesOnVideo(
    videoRef,
    muted,
    volume,
    disableRemotePlayback,
    disablePictureInPicture
  );

  // Get the hover target element from the hoverTarget prop, or default to the component's container div
  const hoverTargetElement = useHoverTargetElement(hoverTarget || containerRef);

  // Keep a ref for the time which the video should be started from next time it is played
  // This is useful if the video gets unloaded and we want to restore it to the time it was
  // at before if the user tries playing it again
  const nextVideoStartTimeRef = useRef(null);

  // Parse the sources and captions into formatted arrays that we can use to
  // render <source> and <track> elements for the video
  const formattedVideoCaptions = useFormatVideoCaptions(videoCaptions);
  const formattedVideoSources = useFormatVideoSrc(
    videoSrc,
    playbackRangeStart,
    playbackRangeEnd
  );
  // Keep a ref to the previous formatted video sources so we can track when the video sources change
  const previousFormattedVideoSourcesRef = useRef<VideoSource[]>(
    formattedVideoSources
  );

  /**
   * Attempts to load and play the video.
   * Storing this on a ref because we don't really want to worry about triggering re-renders when
   * any of this function's dependencies change; it should only be called when the
   * player is hovered/focused.
   */
  const attemptToPlayVideoRef = useRef<() => void>();
  attemptToPlayVideoRef.current = () => {
    const videoElement = videoRef.current;

    if (nextVideoStartTimeRef.current !== null) {
      videoElement.currentTime = nextVideoStartTimeRef.current;
    }

    videoElement.play().catch((error: DOMException) => {
      // If shouldSuppressPlaybackInterruptedErrors is true and this is an AbortError, do nothing instead of logging it.
      if (
        shouldSuppressPlaybackInterruptedErrors &&
        error.name === 'AbortError'
      ) {
        return;
      }

      // Additional handling for when browsers block playback for unmuted videos.
      // This is unfortunately necessary because most modern browsers do not allow playing videos with audio
      //  until the user has "interacted" with the page by clicking somewhere at least once; mouseenter events
      //  don't count.
      // If the video isn't muted and playback failed with a `NotAllowedError`, this means the browser blocked
      // playing the video because the user hasn't clicked anywhere on the page yet.
      if (!videoElement.muted && error.name === 'NotAllowedError') {
        console.warn(
          'HoverVideoPlayer: Playback with sound was blocked by the browser. Attempting to play again with the video muted; audio will be restored if the user clicks on the page.'
        );
        // Mute the video and attempt to play again
        videoElement.muted = true;
        videoElement.play();

        // When the user clicks on the document, unmute the video since we should now
        // be free to play audio
        const onClickDocument = () => {
          videoElement.muted = false;

          // Clean up the event listener so it is only fired once
          document.removeEventListener('click', onClickDocument);
        };
        document.addEventListener('click', onClickDocument);
      } else {
        // Log any other playback errors with console.error
        console.error(`HoverVideoPlayer: ${error.message}`);
      }
    });
  };

  /**
   * Attempts to pause the video.
   * Storing this on a ref because we don't really want to worry about triggering re-renders when
   * any of this function's dependencies change; it should only be called when the
   * player is un-hovered/focused.
   */
  const attemptToPauseVideoRef = useRef<() => void>();
  attemptToPauseVideoRef.current = () => {
    const videoElement = videoRef.current;

    videoElement.pause();

    // Performing post-save cleanup tasks in here rather than the onPause listener
    // because onPause can also be called when the video reaches the end of a playback range
    // and it's just simpler to deal with that separately
    if (restartOnPaused) {
      videoElement.currentTime = playbackRangeStart || 0;
    }
    nextVideoStartTimeRef.current = videoElement.currentTime;

    setIsPlaying(false);
  };

  const [isPlaying, setIsPlaying] = useState(false);
  const [isHovering, setIsHovering] = useState(false);

  useEffect(() => {
    if (!isPlaying) {
      // When the video isn't playing, check if the sources loaded in the video
      // have changed from what the video currently has loaded. If so, we'll call videoElement.load()
      // to trigger a reload with the new source
      const shouldReloadVideoSources =
        previousFormattedVideoSourcesRef.current !== formattedVideoSources;

      if (shouldReloadVideoSources) {
        previousFormattedVideoSourcesRef.current = formattedVideoSources;

        const videoElement = videoRef.current;
        // If the video element doesn't have a loaded source or the source has changed since the
        // last time we played the video, make sure to force the video to load the most up-to-date sources
        videoElement.load();
        // Reset the next start time to the start of the video
        nextVideoStartTimeRef.current = playbackRangeStart || 0;
      }
    }
  }, [formattedVideoSources, isPlaying, playbackRangeStart]);

  const hasPausedOverlay = Boolean(pausedOverlay);
  const hasHoverOverlay = Boolean(hoverOverlay);

  // If we have a paused or hover overlay, the player should wait
  // for the overlay(s) to finish transitioning back in before we
  // pause the video
  const shouldWaitForOverlayTransitionBeforePausing =
    hasPausedOverlay || hasHoverOverlay;

  useEffect(() => {
    if (!hoverTargetElement) return undefined;

    let playbackStartTimeout: number | null = null;
    let pauseTimeout: number | null = null;

    const cancelTimeouts = () => {
      // Cancel any previously active pause or playback attempts
      window.clearTimeout(playbackStartTimeout);
      window.clearTimeout(pauseTimeout);
    };

    const attemptToPlayVideoOnHover = () => {
      cancelTimeouts();

      setIsHovering(true);

      // We only need to attempt to play if the video is currently paused
      if (isVideoElementPaused(videoRef.current)) {
        if (playbackStartDelay) {
          playbackStartTimeout = window.setTimeout(
            () => attemptToPlayVideoRef.current(),
            playbackStartDelay
          );
        } else {
          attemptToPlayVideoRef.current();
        }
      }
    };
    const attemptToPauseVideoOnHoverEnd = () => {
      cancelTimeouts();

      setIsHovering(false);

      if (
        // We only need to delay a pause attempt if the video is currently playing
        isVideoElementPlaying(videoRef.current) &&
        shouldWaitForOverlayTransitionBeforePausing
      ) {
        pauseTimeout = window.setTimeout(
          () => attemptToPauseVideoRef.current(),
          overlayTransitionDuration
        );
      } else {
        attemptToPauseVideoRef.current();
      }
    };

    hoverTargetElement.addEventListener(
      'hvp:hoverStart',
      attemptToPlayVideoOnHover
    );
    hoverTargetElement.addEventListener(
      'hvp:hoverEnd',
      attemptToPauseVideoOnHoverEnd
    );

    return () => {
      hoverTargetElement.removeEventListener(
        'hvp:hoverStart',
        attemptToPlayVideoOnHover
      );
      hoverTargetElement.removeEventListener(
        'hvp:hoverEnd',
        attemptToPauseVideoOnHoverEnd
      );
      cancelTimeouts();
    };
  }, [
    hoverTargetElement,
    overlayTransitionDuration,
    playbackStartDelay,
    shouldWaitForOverlayTransitionBeforePausing,
  ]);

  useManageHoverEvents(
    hoverTargetElement,
    focused,
    disableDefaultEventHandling,
    onHoverStart,
    onHoverEnd
  );

  // We should attempt to play the video if the user is hovering over it or the `focused` override prop is enabled
  // const shouldPlayVideo = isHoveringOverVideo || focused;

  const hasLoadingOverlay = Boolean(loadingOverlay);

  // If the video's sources should be unloaded when it's paused and the video is not currently active, we can unload the video's sources.
  // We will remove the video's <source> tags in this render and then call video.load() in an effect to
  // fully unload the video
  const shouldUnloadVideo = unloadVideoOnPaused && !isHovering && !isPlaying;

  useEffect(() => {
    // If shouldUnloadVideo is true, this effect is being run after the video's
    // sources have been removed, so call load on the video to unload any sources it currently has loaded
    if (shouldUnloadVideo) {
      const videoElement = videoRef.current;
      videoElement.load();
    }
  }, [shouldUnloadVideo]);

  const shouldShowLoadingOverlay = isHovering && !isPlaying;
  // Show a paused overlay when the user isn't hovering or when the user is hovering
  // but the video is still loading
  const shouldShowPausedOverlay = !isHovering || (isHovering && !isPlaying);

  const isUsingPlaybackRange =
    playbackRangeStart !== null || playbackRangeEnd !== null;

  return (
    <div
      data-testid="hover-video-player-container"
      ref={containerRef}
      style={{
        ...containerSizingStyles[sizingMode],
        position: 'relative',
        ...style,
      }}
      {...spreadableProps}
    >
      {hasPausedOverlay ? (
        <div
          style={{
            ...pausedOverlayWrapperSizingStyles[sizingMode],
            zIndex: 1,
            opacity: shouldShowPausedOverlay ? 1 : 0,
            transition: `opacity ${overlayTransitionDuration}ms`,
            // Disable pointer events on the paused overlay when it's hidden
            pointerEvents: shouldShowPausedOverlay ? 'auto' : 'none',
            ...pausedOverlayWrapperStyle,
          }}
          className={pausedOverlayWrapperClassName}
          data-testid="paused-overlay-wrapper"
        >
          {pausedOverlay}
        </div>
      ) : null}
      {hasLoadingOverlay ? (
        <div
          style={{
            ...expandToFillContainerStyle,
            zIndex: 2,
            opacity: shouldShowLoadingOverlay ? 1 : 0,
            transition: `opacity ${overlayTransitionDuration}ms ${
              shouldShowLoadingOverlay ? loadingStateTimeout : 0
            }ms`,
            // Disable pointer events on the loading overlay when it's hidden
            pointerEvents: shouldShowLoadingOverlay ? 'auto' : 'none',
            ...loadingOverlayWrapperStyle,
          }}
          className={loadingOverlayWrapperClassName}
          data-testid="loading-overlay-wrapper"
        >
          {loadingOverlay}
        </div>
      ) : null}
      {hasHoverOverlay ? (
        <div
          style={{
            ...expandToFillContainerStyle,
            zIndex: 3,
            // Show the hover overlay when the player is hovered/playing
            opacity: isHovering ? 1 : 0,
            transition: `opacity ${overlayTransitionDuration}ms`,
            // Disable pointer events on the hover overlay when it's hidden
            pointerEvents: isHovering ? 'auto' : 'none',
            ...hoverOverlayWrapperStyle,
          }}
          className={hoverOverlayWrapperClassName}
          data-testid="hover-overlay-wrapper"
        >
          {hoverOverlay}
        </div>
      ) : null}
      {/* eslint-disable-next-line jsx-a11y/media-has-caption */}
      <video
        // If a playback range is set, the loop attribute will not work correctly so there's no point in setting it here;
        // in that case, we will manually implement this behavior
        loop={isUsingPlaybackRange ? false : loop}
        playsInline
        preload={preload}
        crossOrigin={crossOrigin}
        ref={videoRef}
        style={{
          ...videoSizingStyles[sizingMode],
          objectFit: 'cover',
          ...videoStyle,
        }}
        controls={controls}
        controlsList={controlsList}
        className={videoClassName}
        id={videoId}
        data-testid="video-element"
        onPlaying={() => {
          setIsPlaying(true);
        }}
        onTimeUpdate={
          // If there's a playback range set, the traditional `loop` video prop won't work correctly so
          // we'll need watch the video's time as it plays and manually keep it within the bounds of the range
          isUsingPlaybackRange
            ? () => {
                const videoElement = videoRef.current;

                const maxVideoTime = playbackRangeEnd || videoElement.duration;
                const minVideoTime = playbackRangeStart || 0;

                const { currentTime } = videoElement;

                if (loop && currentTime >= maxVideoTime) {
                  // If the video should loop and is >= the max video time,
                  // loop it back around to the start
                  const startTime = playbackRangeStart || 0;
                  videoElement.currentTime = startTime;

                  // If the video is paused but the user is still hovering,
                  // meaning it should continue to play, call play() to keep it going
                  if (isHovering && isVideoElementPaused(videoElement)) {
                    videoElement.play();
                  }
                } else if (currentTime > maxVideoTime) {
                  // If the video shouldn't loop but we've exceeded the max video time,
                  // clamp it to the max time and pause it
                  videoElement.pause();
                  videoElement.currentTime = maxVideoTime;
                } else if (currentTime < minVideoTime) {
                  // If the video's time somehow ended up before the min video time,
                  // clamp it to the min time
                  videoElement.currentTime = minVideoTime;
                }
              }
            : null
        }
      >
        {!shouldUnloadVideo &&
          // Only render sources for the video if it is not unloaded
          formattedVideoSources.map(({ src, type }) => (
            <source key={src} src={src} type={type} />
          ))}
        {formattedVideoCaptions.map(
          ({ src, srcLang, label, kind, default: isDefault }) => (
            <track
              key={src}
              kind={kind}
              src={src}
              srcLang={srcLang}
              label={label}
              default={isDefault}
            />
          )
        )}
      </video>
    </div>
  );
}
