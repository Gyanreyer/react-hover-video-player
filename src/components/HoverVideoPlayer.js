import React from 'react';
import { css, cx } from 'emotion';

import FadeTransition from './FadeTransition';

// Enumerates states that the hover player can be in
const HOVER_PLAYER_STATE = {
  stopped: 0,
  stopping: 1,
  loading: 2,
  playing: 3,
};

const expandToCoverVideo = css`
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
`;

/**
 * @typedef   {object}  VideoSource
 * @property  {string}  src - The src URL string to use for a video player source
 * @property  {string}  type - The media type of the video, ie 'video/mp4'
 */

/**
 * @typedef   {object}  VideoCaptionsTrack
 * @property  {string}  src - The src URL string for the captions track file
 * @property  {string}  srcLang - The language code for the language that these captions are in
 * @property  {string}  label - The title of the captions track
 */

/**
 * @component HoverVideoPlayer
 *
 * @param {!(string|string[]|VideoSource|VideoSource[])}  videoSrc - Source(s) to use for the video player. Accepts 3 different formats:
 *                                                                   - **String**: the URL string to use as the video player's src
 *                                                                   - **Object**: an object with attributes:
 *                                                                     - src: The src URL string to use for a video player source
 *                                                                     - type: The media type of the video source, ie 'video/mp4'
 *                                                                   - **Array**: if you would like to provide multiple sources, you can provide an array of URL strings and/or objects with the shape described above
 * @param {!(string|string[]|VideoCaptionsTrack|VideoCaptionsTrack[])} videoCaptions - Captions track(s) to use for the video player for accessibility. Accepts 3 different formats:
 *                                                                                     - **String**: the URL string to use as the captions track's src
 *                                                                                     - **Object**: an object with attributes:
 *                                                                                       - src: The src URL string for the captions track file
 *                                                                                       - srcLang: The language code for the language that these captions are in (ie, 'en', 'es', 'fr')
 *                                                                                       - label: The title of the captions track
 *                                                                                     - **Array**: if you would like to provide multiple caption tracks, you can provide an array of objects with the shape described above
 * @param {bool}    [isFocused=false] - Offers a prop interface for forcing the video to start/stop without DOM events
 *                                        When set to true, the video will begin playing and any events that would normally
 *                                        stop it will be ignored
 * @param {node}    [pausedOverlay] - Contents to render over the video while it's not playing
 * @param {node}    [loadingOverlay] - Contents to render over the video while it's loading
 * @param {number}  [overlayFadeTransitionDuration=400] - The transition duration in ms for how long it should take for the overlay to fade in/out
 * @param {bool}    [shouldRestartOnVideoStopped=true] - Whether the video should reset to the beginning every time it stops playing after the user mouses out of the player
 * @param {bool}    [shouldUseOverlayDimensions=true] - Whether the player should display using the pausedOverlay's dimensions rather than the video element's dimensions if an overlay is provided
 *                                                        This helps prevent content height jumps when the video loads its dimensions
 * @param {func}    [onStartingVideo] - Optional callback for every time the user mouses over or focuses on the hover player and we attempt to start the video
 * @param {func}    [onStartedVideo] - Optional callback for when the video has been successfully started
 * @param {func}    [onStoppingVideo] - Optional callback for every time the user mouses out or blurs the hover player and we attempt to stop the video
 * @param {func}    [onStoppedVideo] - Optional callback for when the video has successfully been stopped
 * @param {bool}    [isVideoMuted=true] - Whether the video player should be muted
 * @param {bool}    [shouldShowVideoControls=false] - Whether the video player should show the browser's controls
 * @param {bool}    [shouldVideoLoop=true] - Whether the video player should loop when it reaches the end
 * @param {string}  [className] - Optional className to apply custom styling to the container element
 * @param {string}  [pausedOverlayWrapperClassName] - Optional className to apply custom styling to the overlay contents' wrapper
 * @param {string}  [loadingOverlayWrapperClassName] - Optional className to apply custom styling to the loading state overlay contents' wrapper
 * @param {string}  [videoClassName] - Optional className to apply custom styling to the video element
 * @param {object}  [style] - Style object to apply custom CSS styles to the hover player container
 *
 * @license MIT
 */
function HoverVideoPlayer({
  videoSrc,
  videoCaptions,
  isFocused = false,
  pausedOverlay = null,
  loadingOverlay = null,
  overlayFadeTransitionDuration = 400,
  shouldRestartOnVideoStopped = true,
  shouldUseOverlayDimensions = true,
  onStartingVideo = null,
  onStartedVideo = null,
  onStoppingVideo = null,
  onStoppedVideo = null,
  isVideoMuted = true,
  shouldShowVideoControls = false,
  shouldVideoLoop = true,
  className = '',
  pausedOverlayWrapperClassName = '',
  loadingOverlayWrapperClassName = '',
  videoClassName = '',
  style = null,
}) {
  const [hoverPlayerState, setHoverPlayerState] = React.useState(
    HOVER_PLAYER_STATE.stopped
  );

  const containerRef = React.useRef();
  const videoRef = React.useRef();
  const pauseVideoTimeoutRef = React.useRef();

  // Parse the `videoSrc` prop into an array of VideoSource objects to be used for the video player
  const parsedVideoSources = React.useMemo(() => {
    if (videoSrc == null) {
      // A videoSrc value is required in order to make the video player work
      console.error(
        "Error: 'videoSrc' prop is required for HoverVideoPlayer component"
      );

      return [];
    }

    return (
      // Make sure we can treat the videoSrc value as an array
      []
        .concat(videoSrc)
        // Parse our video source values into an array of VideoSource objects that can be used to render sources for the video
        .reduce((sourceArray, source) => {
          if (typeof source === 'string') {
            // If the source is a string, it's an src URL so format it into a VideoSource object and add it to the array
            sourceArray.push({ src: source });
          } else if (source && source.src) {
            // If the source is an object with an src, just add it to the array
            sourceArray.push({ src: source.src, type: source.type });
          } else {
            // Log an error if one of the videoSrc values is invalid
            console.error(
              "Error: invalid value provided to HoverVideoPlayer prop 'videoSrc':",
              source
            );
          }

          return sourceArray;
        }, [])
    );
  }, [videoSrc]);

  const parsedVideoCaptions = React.useMemo(() => {
    // If no captions were provided, return an empty array
    if (videoCaptions == null) return [];

    return (
      // Make sure we can treat the videoSrc value as an array
      []
        .concat(videoCaptions)
        // Parse our video captions values into an array of VideoCaptionsTrack
        // objects that can be used to render caption tracks for the video
        .reduce((captionsArray, captions) => {
          if (typeof captions === 'string') {
            captionsArray.push({ src: captions });
          } else if (captions && captions.src) {
            captionsArray.push({
              src: captions.src,
              srcLang: captions.srcLang,
              label: captions.label,
            });
          } else {
            // Log an error if one of the videoCaptions values is invalid
            console.error(
              "Error: invalid value provided to HoverVideoPlayer prop 'videoCaptions'",
              captions
            );
          }

          return captionsArray;
        }, [])
    );
  }, [videoCaptions]);

  /**
   * @function  attemptStartVideo
   *
   * Starts the video and fades the paused overlay out when the user mouses over or focuses in the video container element
   */
  function attemptStartVideo() {
    // If the user quickly moved their mouse away and then back over the container,
    // cancel any outstanding timeout that would pause the video
    clearTimeout(pauseVideoTimeoutRef.current);

    // If the video is already loading/playing, return early
    if (hoverPlayerState >= HOVER_PLAYER_STATE.loading) return;

    // If we have an onStartingVideo callback, fire it to indicate the video is attempting to start
    if (onStartingVideo) onStartingVideo();

    if (videoRef.current.readyState >= 3) {
      // If the video's readyState indicates that it is already loaded enough to begin playing,
      // skip the loading state
      setHoverPlayerState(HOVER_PLAYER_STATE.playing);
    } else {
      // Otherwise if the video is not loaded enough to play yet, show a loading state until the
      // `onPlaying` event is fired
      setHoverPlayerState(HOVER_PLAYER_STATE.loading);
    }

    if (videoRef.current.paused) {
      // If the video is not playing and has no play attempt in progress, start a play attempt
      videoRef.current.play();
    }
  }

  /**
   * @function  attemptStopVideo
   *
   * Stops the video and fades the paused overlay in when the user mouses out from or blurs the video container element
   */
  const attemptStopVideo = React.useCallback(() => {
    // If the isFocused override prop is active, ignore any other events attempting to stop the video
    if (isFocused) return;

    // Return early if the video is already stopped
    if (hoverPlayerState <= HOVER_PLAYER_STATE.stopping) return;

    // If we have an onStoppingVideo callback, fire it to indicate the video is in the process of being stopped
    if (onStoppingVideo) onStoppingVideo();

    // Start fading the overlay back in to cover up the video before it's paused
    setHoverPlayerState(HOVER_PLAYER_STATE.stopping);

    // If we already had a timeout going to pause the video, cancel it so we can
    // replace it with a new one
    clearTimeout(pauseVideoTimeoutRef.current);

    // Set a timeout with the duration of the overlay's fade transition so the video
    // won't stop until it's fully hidden
    pauseVideoTimeoutRef.current = setTimeout(
      () => {
        if (hoverPlayerState === HOVER_PLAYER_STATE.playing) {
          // If the video is playing, pause it
          videoRef.current.pause();
        }
      },
      // If we don't have to wait for the overlay to fade back in, just set the timeout to 0 to invoke it ASAP
      pausedOverlay ? overlayFadeTransitionDuration : 0
    );
  }, [
    hoverPlayerState,
    isFocused,
    onStoppingVideo,
    overlayFadeTransitionDuration,
    pausedOverlay,
  ]);

  React.useEffect(() => {
    // Manually setting the `muted` attribute on the video element via an effect in order
    // to avoid a know React issue with the `muted` prop not applying correctly on initial render
    // https://github.com/facebook/react/issues/10389
    videoRef.current.muted = isVideoMuted;
  }, [isVideoMuted]);

  React.useEffect(() => {
    // Use effect to start/stop the video when isFocused override prop changes
    if (isFocused) {
      attemptStartVideo();
    } else {
      attemptStopVideo();
    }

    // We really only want to fire this effect when the isFocused prop changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isFocused]);

  React.useEffect(() => {
    const onWindowTouchStart = (event) => {
      if (!containerRef.current.contains(event.target)) {
        attemptStopVideo();
      }
    };

    window.addEventListener('touchstart', onWindowTouchStart);

    // Remove the event listener on cleanup
    return () => window.removeEventListener('touchstart', onWindowTouchStart);
  }, [attemptStopVideo]);

  const shouldVideoExpandToFitOverlayDimensions =
    pausedOverlay && shouldUseOverlayDimensions;

  return (
    <div
      onMouseEnter={attemptStartVideo}
      onFocus={attemptStartVideo}
      onMouseOut={attemptStopVideo}
      onBlur={attemptStopVideo}
      onTouchStart={attemptStartVideo}
      className={cx(
        css`
          position: relative;
        `,
        className
      )}
      style={style}
      data-testid="hover-video-player-container"
      ref={containerRef}
    >
      {pausedOverlay && (
        <FadeTransition
          isVisible={hoverPlayerState <= HOVER_PLAYER_STATE.loading}
          duration={overlayFadeTransitionDuration}
          className={cx(
            css`
              position: relative;
              display: block;
              width: 100%;
              z-index: 1;
              pointer-events: none;

              & > * {
                display: block;
              }
            `,
            {
              [expandToCoverVideo]: !shouldVideoExpandToFitOverlayDimensions,
            },
            pausedOverlayWrapperClassName
          )}
          testID="paused-overlay-transition-wrapper"
        >
          {pausedOverlay}
        </FadeTransition>
      )}
      {loadingOverlay && (
        <FadeTransition
          isVisible={hoverPlayerState === HOVER_PLAYER_STATE.loading}
          duration={overlayFadeTransitionDuration}
          shouldMountOnEnter
          className={cx(
            css`
              z-index: 2;
              pointer-events: none;
            `,
            expandToCoverVideo,
            loadingOverlayWrapperClassName
          )}
          testID="loading-overlay-transition-wrapper"
        >
          {loadingOverlay}
        </FadeTransition>
      )}
      {/* eslint-disable-next-line jsx-a11y/media-has-caption */}
      <video
        controls={shouldShowVideoControls}
        loop={shouldVideoLoop}
        playsInline
        // Only preload video data if we depend on having loaded its dimensions to display it
        preload={shouldVideoExpandToFitOverlayDimensions ? 'none' : 'metadata'}
        ref={videoRef}
        className={cx(
          css`
            width: 100%;
            display: block;
            object-fit: cover;
          `,
          {
            [expandToCoverVideo]: shouldVideoExpandToFitOverlayDimensions,
          },
          videoClassName
        )}
        onPlaying={() => {
          if (hoverPlayerState >= HOVER_PLAYER_STATE.loading) {
            // If the player was showing a loading state, transition into the playing state
            setHoverPlayerState(HOVER_PLAYER_STATE.playing);

            // If we have an onStartedVideo callback, fire it to indicate the video has successfully started
            if (onStartedVideo) onStartedVideo();
          } else {
            // If our play attempt completed but the user has since moused away and is not looking
            // at the loading state, immediately pause the video until the user comes back
            videoRef.current.pause();
          }
        }}
        onPause={() => {
          // On the off chance that this event was fired but there is still an outstanding pause timeout, cancel it
          clearTimeout(pauseVideoTimeoutRef.current);

          if (hoverPlayerState >= HOVER_PLAYER_STATE.stopping) {
            // If the player was in a loading/playing state, transition into the stopped state
            // Mark that the video has been stopped
            setHoverPlayerState(HOVER_PLAYER_STATE.stopped);

            // If we have an onStoppedVideo callback, fire it to indicate the video has been stopped
            if (onStoppedVideo) onStoppedVideo();
          }

          if (shouldRestartOnVideoStopped) {
            // If we should restart the video, reset its time to the beginning
            videoRef.current.currentTime = 0;
          }
        }}
        onError={() => {
          setHoverPlayerState(HOVER_PLAYER_STATE.stopped);
        }}
      >
        {parsedVideoSources.map(({ src, type }) => (
          <source key={src} src={src} type={type} />
        ))}
        {parsedVideoCaptions.map(({ src, srcLang, label }) => (
          <track
            key={src}
            kind="captions"
            src={src}
            srcLang={srcLang}
            label={label}
          />
        ))}
      </video>
    </div>
  );
}

export default HoverVideoPlayer;
