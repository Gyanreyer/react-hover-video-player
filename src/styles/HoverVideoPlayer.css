.Container {
  position: relative;
}

.PausedOverlayContainer {
  position: relative;
  display: block;
  width: 100%;
  z-index: 1;
  pointer-events: none;
}

.LoadingOverlayContainer {
  z-index: 2;
  pointer-events: none;
}

.Video {
  width: 100%;
  display: block;
}

.ExpandToCoverContainer {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
}

.ShouldExpandToCoverVideo, .LoadingOverlayContainer, .ShouldExpandToFitOverlay {
  composes: ExpandToCoverContainer;
}

.PausedOverlayContainer > * {
  display: block;
}

.Video.ShouldExpandToFitOverlay {
  object-fit: cover;
}