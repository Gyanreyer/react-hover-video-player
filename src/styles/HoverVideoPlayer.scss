.Container {
  position: relative;
}

.PausedOverlayContainer, .LoadingOverlayContainer {
  /* Overlay wrapper should completely cover the dimensions of the video */
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
  pointer-events: none;
}

.LoadingOverlayContainer {
  z-index: 1;
}

.Video {
  width: 100%;
  display: block;
}