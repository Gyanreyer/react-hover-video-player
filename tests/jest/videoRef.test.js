import React from 'react';
import { render, screen } from '@testing-library/react';

// This is being mapped in the jest config files so we can alternate between
// running tests off of the development version (../src) of the component or
// the built production version (../es)
// eslint-disable-next-line import/no-unresolved
import HoverVideoPlayer from 'react-hover-video-player';

describe('videoRef', () => {
  afterEach(() => {
    // No errors should be logged in any of these tests
    expect(console.error).not.toHaveBeenCalled();
  });

  test("correctly forwards HoverVideoPlayer's video element to the videoRef prop", () => {
    let videoRefValue;

    function PlayerWithVideoRef() {
      const videoRef = React.useRef();

      React.useEffect(() => {
        videoRefValue = videoRef.current;
      }, []);

      return <HoverVideoPlayer videoSrc="fake-video.mp4" videoRef={videoRef} />;
    }

    const { container } = render(<PlayerWithVideoRef />);
    expect(container).toMatchSnapshot();

    expect(videoRefValue).toBe(screen.getByTestId('video-element'));
  });
});
