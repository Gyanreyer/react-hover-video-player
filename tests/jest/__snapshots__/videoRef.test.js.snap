// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`videoRef correctly forwards HoverVideoPlayer's video element to the videoRef prop 1`] = `
<div>
  <div
    data-testid="hover-video-player-container"
    style="position: relative;"
  >
    <video
      crossorigin="anonymous"
      data-testid="video-element"
      loop=""
      playsinline=""
      style="display: block; width: 100%; object-fit: cover;"
    >
      <source
        src="fake-video.mp4"
      />
    </video>
  </div>
</div>
`;
